tests/tests.rs

```diff
@@ -703,6 +710,131 @@ async fn test_unix_socket() {
+    S::Future: Send,
+{
+    let port = PORT.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+    let tcp_listener = tokio::net::TcpListener::bind(("127.0.0.1", port))
+        .await
+        .unwrap();
+
+    let cert = make_root_cert();
+    let proxy = Arc::new(http_mitm_proxy::MitmProxy::new(Some(cert), None));
+
+    let serve = async move {
+        loop {
+            let (stream, _) = tcp_listener.accept().await.unwrap();
+
+            let proxy = proxy.clone();
+            let service = service.clone();
+            tokio::spawn(async move {
+                if http2 {
+                    let _ = hyper::server::conn::http2::Builder::new(TokioExecutor::new())
+                        .serve_connection(
+                            TokioIo::new(stream),
+                            MitmProxy::wrap_service(proxy, service),
+                        )
+                        .await;
+                } else {
+                    let _ = hyper::server::conn::http1::Builder::new()
+                        .preserve_header_case(true)
+                        .title_case_headers(true)
+                        .serve_connection(
+                            TokioIo::new(stream),
+                            MitmProxy::wrap_service(proxy, service),
+                        )
+                        .with_upgrades()
+                        .await;
+                }
+            });
+        }
+    };
+
+    (port, serve)
+}
+
+async fn test_proxy_with_setting(https: bool, http2: bool, proxy_http2: bool) {
+    let (proxy_port, proxy_serve) = bind_proxy(
+        service_fn(|_req| async {
+            let res = Response::new("Hello World".to_string());
+            Ok::<_, Infallible>(res)
+        }),
+        proxy_http2,
+    )
+    .await;
+
+    tokio::spawn(proxy_serve);
+
+    let cargo_bin = Command::cargo_bin("oha").unwrap();
+    let mut proc = tokio::process::Command::new(cargo_bin.get_program());
+    std::mem::drop(cargo_bin);
+
+    let scheme = if https { "https" } else { "http" };
+    proc.args(["--no-tui", "--debug", "--insecure", "-x"])
+        .arg(format!("http://127.0.0.1:{proxy_port}/"))
+        .arg(format!("{scheme}://example.com/"));
+    if http2 {
+        proc.arg("--http2");
+    }
+    if proxy_http2 {
+        proc.arg("--proxy-http2");
+    }
+
+    proc.stdin(std::process::Stdio::null())
+        .stdout(std::process::Stdio::piped())
+        .stderr(std::process::Stdio::null());
+    let stdout = proc
+        .spawn()
+        .unwrap()
+        .wait_with_output()
+        .await
+        .unwrap()
+        .stdout;
+
+    assert!(String::from_utf8(stdout).unwrap().contains("Hello World"),);
+}
+
+#[tokio::test]
+async fn test_proxy() {
+    for https in [false, true] {
+        for http2 in [false, true] {
+            for proxy_http2 in [false, true] {
+                if https && proxy_http2 {
+                    // Waiting for https://github.com/rustls/tokio-rustls/pull/93 is released
```

Comment:
This test will be unlocked when https://github.com/rustls/tokio-rustls/pull/93 has released